Types cdata wrapping

exvito opened the pull request:
@opalmer 

Here is the first take at the PR. General notes:
* I followed the strategy generally outlined in the mailing list: https://groups.google.com/forum/#!topic/pywincffi/B9JNULyvhgI
* I did lots of small commits, in what I believe is a mostly sane order (thus, to better understand the PR, it may be useful to look at a few commits at a time).
* General result:
  * The end user APIs no longer expose CFFI <cdata> objects: whether arguments or results.
  * Full separation/clarification of what is text and what is binary at the API wrapper level.
* Details:
  * C structs are now wrapped in simple, light but customizable Python objects.
  * HANDLEs are also wrapped; they're even lighter than structs, and need special handling, but I thing the API utilization results in a much nicer experice.
  * Type validation is simplified and the input_check function reflects that.
  * Given strict text/binary separation string_to_cdata no longer needed.
  * Some tests had to be adjusted to the revamped types / APIs.
  * Some no longer made sense and I threw them away.
* Doc strings will need to be reviewed.
* Tests are passing here.
* There will certainly things that need fixing, of course.

Your input is welcome.
Cheers.

exvito commented:
Note:
- Some tests are failing on Python 3.x -- I'll go ahead and address that next. (no Py 3 environment here!)
- I'm not sure how to "fix" some of the "PYLINT=1" complaints -- I'll try and improve on that as well.
- The "READTHEDOCS=1" failure looks like a mystery -- I'll leave that for last.

I forgot to mention:
- I used one cffi 1.6.0 API (having had to upgrade my venv) in kernel32/file.py on the last return statement of ReadFile: its ffi.unpack as documented here http://cffi.readthedocs.io/en/latest/ref.html#ffi-string-ffi-unpack

exvito commented:
Further notes:
- I guess I finally fixed the Python 3 tests and pylint errors/warnings.
- What I need your expertise for: addressing the READTHEDOCS=1 failures which involve tech I'm not familiar with at all (travis + sphinx). I find it odd, that 2.x and 3.x versions seem to be failing differently.

Waiting on your input.

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+
+# pylint: disable=too-few-public-methods
+class CFFICDataWrapper(object):
+
```
Shouldn't have newlines before the doc string starts.

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+    * Wraps a CFFI cdata object in self._cdata.
+    * Delegates attribute getting/setting to self._cdata, supporting structs.
+    * Delegates item getting/setting to self._cdata, supporting arrays.
+    Attribute access is not delegated to the wrapped object if the class
```
nit: newline before the paragraph starts

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+class CFFICDataWrapper(object):
+
+    """
+    Base class for exposing Python types and interfaces to pywincffi users.
```
nit: newline before the bullet list starts, probably would be nice to have : at the end of this line too.

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+
+    def __init__(self, cdecl, ffi=None):
+        if ffi is None:
+            ffi = cffi.FFI()
```
You'll want to use `dist.load()` here to retrieve the ffi object:

``` python
if ffi is None:
    ffi, _ = dist.load()
```

dist.py caches an instance of `cffi.FFI` which may be loaded from the precompiled version of pywincffi.

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+
+    def __setattr__(self, name, value):
+        # avoid self-recursion setting own attribute: use parent's __setattr__
+        if name == '_cdata':
```
' -> " (for consistency)

opalmer commented on pywincffi/core/typesbase.py:
```diff
@@ -0,0 +1,58 @@
+    * Delegates item getting/setting to self._cdata, supporting arrays.
+    Attribute access is not delegated to the wrapped object if the class
+    itself contains such an attribute and that attribute is a descriptor; this
+    is in place to support @property in sub-classes.
```
Could you add some parameter and keyword documentation for cdecl and ffi?  Does not have to be anything complex, just enough to describe the inputs.

exvito replied:
Making ffi a non-optional arg. Benefit: Ensure loose coupling between this class and the remaining code.

Notes:
- wintypes was already providing the dist.load() ffi instance.
- test_typesbase with minor review, explicitly passing "dummy" ffi instance.

opalmer commented on pywincffi/kernel32/events.py:
```diff
@@ -34,7 +34,8 @@ def CreateEvent(
         then by default the handle cannot be inherited by a subprocess.
 
-    :keyword str lpName:
+    :keyword unicode/str lpName:
```
I was going to comment on this and suggest:

```
:type lpName: unicode or str
:keyword lpName:
```

or 

```
:keyword str lpName:
```

After playing around with it some though, I don't think it matters (so this is more of an FYI I guess).  If you put just `str` in there sphinx will create a link to the standard string type.  It seems it allows `unicode/str` and the `:type ...` form I have above but neither of those create links back for the standard types.

opalmer commented on pywincffi/kernel32/events.py:
```diff
@@ -88,24 +86,25 @@ def OpenEvent(dwDesiredAccess, bInheritHandle, lpName):
     :param bool bInheritHandle:
-    :param str lpName:
+    :param unicode/str lpName:
+        Type is unicode on Python 2, str on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/kernel32/events.py:
```diff
@@ -34,7 +34,8 @@ def CreateEvent(
 
-    :keyword str lpName:
+    :keyword unicode/str lpName:
+        Type is unicode on Python 2, str on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/kernel32/file.py:
```diff
@@ -27,7 +29,8 @@ def CreateFile(  # pylint: disable=too-many-arguments
 
-    :param str lpFileName:
+    :param unicode/str lpFileName:
+        Type is unicode on Python 2, str on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/kernel32/file.py:
```diff
@@ -131,66 +125,53 @@ def WriteFile(hFile, lpBuffer, nNumberOfBytesToWrite=None, lpOverlapped=None,
     :param lpBuffer:
-        The data to be written to the file or device. We should be able
-        to convert this value to unicode.
+        Type is str on Python 2, bytes on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/kernel32/file.py:
```diff
@@ -131,66 +125,53 @@ def WriteFile(hFile, lpBuffer, nNumberOfBytesToWrite=None, lpOverlapped=None,
+        >>> from pywincffi.kernel32 import WriteFile, CreateEvent
+        >>> from pywincffi.wintypes import OVERLAPPED
+        >>> hEvent = CreateEvent(...)
+        >>> lpOverlapped = OVERLAPPED()
```
Thanks for updating the example both here and below 👍

opalmer commented on pywincffi/kernel32/file.py:
```diff
@@ -274,10 +254,12 @@ def MoveFileEx(lpExistingFileName, lpNewFileName, dwFlags=None):
 
-    :param str lpExistingFileName:
+    :param unicode/str lpExistingFileName:
+        Type is unicode on Python 2, str on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/kernel32/file.py:
```diff
@@ -274,10 +254,12 @@ def MoveFileEx(lpExistingFileName, lpNewFileName, dwFlags=None):
 
-    :param str lpNewFileName:
+    :param unicode/str lpNewFileName:
+        Type is unicode on Python 2, str on Python 3.
```
I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+"""
+Windows Types Package
+====================
```
The footer (===) needs to be as long as the section header:

```
"""
Windows Types Package
=====================

Provides user accessible types corresponding to the respective Windows types
used across the exposed APIs.
"""
```

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+from pywincffi.core import dist
+
+
+_FFI, _ = dist.load()
```
So this is why the doc build steps fail.

When you call `dist.load()` it will either import `_pywincffi` or try to build it if it does not exist.  Travis builds the documentation on Linux just like [readthedocs.org](https://readthedocs.org/projects/pywincffi/builds/) across multiple Python versions so if there's a problem like this it's easy to discover while doing development versus after the fact.  Of course Linux does not have access the headers and libraries that Windows does so this in turn causes the build to fail.

I suggest using `dist.load()` inside of each of the functions in this module like we do elsewhere in pywincffi.  This prevents pywincffi from being built and allows static tools like sphinx and pylint to run on any platform.

exvito replied:
Thanks for clarifying. Changing this.

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+    """
+    :param wintype:
+        A type derived from :class:`pywincffi.core.typesbase.CFFICDataWrapper`
+    :return:
```
Newline before `:return:`

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+# pylint: disable=protected-access
+def wintype_to_cdata(wintype):
+    """
+    :param wintype:
```
Could we get a brief description of the function's purpose and maybe a small example?

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+# pylint: disable=too-few-public-methods
+class HANDLE(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/aa383751
```
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/aa383751
```

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+
+    def __eq__(self, other):
+        if not isinstance(other, HANDLE):
+            raise TypeError('%r must be a HANDLE' % other)
```
' -> " (for consistency)

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+# pylint: disable=invalid-name
+class SECURITY_ATTRIBUTES(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/aa379560
```
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/aa379560
```

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+# pylint: disable=too-few-public-methods
+class OVERLAPPED(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/ms684342
```
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/ms684342
```

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+    @hEvent.setter
+    def hEvent(self, handle):
+        if not isinstance(handle, HANDLE):
+            raise TypeError('%r must be a HANDLE' % handle)
```
' -> " (for consistency)

opalmer commented on pywincffi/wintypes/__init__.py:
```diff
@@ -0,0 +1,113 @@
+# pylint: disable=too-few-public-methods
+class FILETIME(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/ms724284
```
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/ms724284
```

opalmer commented:
Did a review and overall it looks great, thanks again for putting the time into this.  Based on the changes in the tests around input/output type handling this definitely looks like it will be a major improvement once it lands.

Most of my comments where around consistency, doc strings, etc so nothing major to really fix.  I plan to do a deeper dive and play around with the new types some myself before a final review but I don't think that will end up changing anything.  

So there's really only two fixes here that break things and need to be fixed:
- You need to call dist.load() inside of the functions in typebases.py
- Some doc string formatting which could result in odd rendering issues.
- Update the cffi requirement in setup.py to `cffi>=1.6.0`.  As you stated in your comment above we're using `ffi.unpack` which came out in 1.6.0.  Since we rely on that API we have to make sure downstream consumers of pywincffi are using the proper cffi version.

exvito commented:
I believe I addressed all of your suggestions with the prior commit.
Other than any further review you may share, I think the following actions are still pending:
- Removing the pywincffi.util module - string_to_cdata no longer used.
- Reviewing docstrings in the wrappers.
- Updating the change log.
- Not sure - Maybe the docs need review?

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -4,12 +4,15 @@
     """
     def test_instantiate(self):
-        o = typesbase.CFFICDataWrapper("char *")
+        o = typesbase.CFFICDataWrapper("char *", _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -19,7 +22,7 @@ def test_simple_struct_set_and_get(self):
                 float z;
             } *
-        """)
+        """, _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -28,7 +31,7 @@ def test_simple_struct_set_and_get(self):
 
     def test_char_array_set_and_get(self):
-        o = typesbase.CFFICDataWrapper("char [256]")
+        o = typesbase.CFFICDataWrapper("char [256]", _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -43,16 +46,15 @@ def test_char_array_set_and_get(self):
     def test_char_array_set_and_ffi_string(self):
-        ffi = cffi.FFI()
-        o = typesbase.CFFICDataWrapper("char [256]")
+        o = typesbase.CFFICDataWrapper("char [256]", _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
                 } u;
             } *
-        """)
+        """, _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
         _file_name = b"the-filename.test"
         o.u.file.name = _file_name
-        self.assertEqual(ffi.unpack(o.uuid, len(_uuid)), _uuid)
+        self.assertEqual(_ffi.unpack(o.uuid, len(_uuid)), _uuid)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -61,7 +63,7 @@ def test_simple_struct_array_set_and_get(self):
                 int second;
             } [20]
-        """)
+        """, _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -43,16 +46,15 @@ def test_char_array_set_and_get(self):
         # must use wrapped cdata to apply ffi.string in this case
         # won't be needed for struct members as other tests verify
-        self.assertEqual(ffi.string(o._cdata), content)
+        self.assertEqual(_ffi.string(o._cdata), content)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
         self.assertEqual(o.u.file.inode, 12345)
         self.assertEqual(
-            ffi.unpack(o.u.file.name, len(_file_name)),
+            _ffi.unpack(o.u.file.name, len(_file_name)),
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
     def test_externally_provided_ffi(self):
-        ffi = cffi.FFI()
-        ffi.cdef("""
+        _ffi.cdef("""
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
             } person_t;
-        """)
-        o = typesbase.CFFICDataWrapper("person_t *", ffi)
+        """, _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
-        """)
-        o = typesbase.CFFICDataWrapper("person_t *", ffi)
+        """, _ffi)
+        o = typesbase.CFFICDataWrapper("person_t *", _ffi)
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
         o.last_name = u"Last Name"
-        self.assertEqual(ffi.string(o.first_name), u"First Name")
-        self.assertEqual(ffi.string(o.last_name), u"Last Name")
+        self.assertEqual(_ffi.string(o.first_name), u"First Name")
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented on tests/test_core/test_typesbase.py:
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
-        self.assertEqual(ffi.string(o.first_name), u"First Name")
-        self.assertEqual(ffi.string(o.last_name), u"Last Name")
+        self.assertEqual(_ffi.string(o.first_name), u"First Name")
+        self.assertEqual(_ffi.string(o.last_name), u"Last Name")
```
Should use dist.load() to retrieve an instance of ffi like we do in normal code.

opalmer commented:
Changes look good, thanks.  Had a few comments on the tests though but it's basically all the same comment: we should use dist.load() in the tests like we do in normal code.

I also agree with your list of TODOs, especially the doc string review (which is something I intend to take another pass at myself too).

exvito commented:
My idea is to keep pywincffi/core/types.base (CFFICDataWrapper) + associated tests (test_typesbase.py you commented on above) completely uncoupled from the rest of the code. That's why those tests use independent FFI instances. In other words: the class and associated tests have zero references/dependencies on the remaining code, they should only depend on CFFI.

On the other hand, the actual types in use (pywincffi/wintypes/**init**.py) do have their tests dependent on the project code and, in particular, in the FFI object obtained from dist.load() in tests/test_wintypes. Those are "implementing" C types declared in the header files that dist.load() handles and need to be tested against that.

If you don't agree with this approach, I can change that, of course. But I think keeping the CFFICDataWrapper + tests with no dependencies is a cleaner design with no drawbacks that I can identify now, while future-proofing against future changes (which will be simpler with less dependencies).

Let me know of your thoughts.

opalmer commented:
Alright that seems fine to me, thanks for explaining. I took another look at the docs and unless you have further changes I think I'd be ok with merging this.

Thanks again! 👍

exvito commented:
Completed:
- Reviewed doc strings and updated types.
- Removed the no longer used pywincffi/util.py containing the string_to_cdata function.
- Updated changelog.rst.

For some reason not clear to me, the tests here are failing complaining that this branch has conflicts. Why it won't run the tests I don't understand (even though there may be merge conflicts, of course).

Other than that, I agree with you, and if tests come out ok, I think this is ready to be merged.

exvito commented:
Investigated a bit and found out that the merge conflict came down to a simple thing in docs/source/changelog.rst. Merged in your master, fixed changelog.rst manually and away it goes.

Let's see if the tests come out ok for travis and appveyor.

opalmer commented:
Looks good, not seeing anything additional here that needs to be fixed so I'll go ahead and merge this.  This also fixes #86.  Thanks again, looking forward to using these changes in future PRs!
